sdk/storage/AzureStorageBlob/Source/DataStructures/StorageStreamUploader.swift

```diff
@@ -0,0 +1,540 @@
+        fatalError("Client-side encryption is not currently supported!")
+    }
+}
+
+public struct BlobUploadProgress: TransferProgress {
+    public var bytes: Int
+    public var totalBytes: Int
+
+    public var asPercent: Int {
+        return Int(asFloat * 100.0)
+    }
+
+    public var asFloat: Float {
+        return Float(bytes) / Float(totalBytes)
+    }
+}
+
+public protocol BlobUploadDelegate: AnyObject {
+    func uploader(
+        _ uploader: BlobStreamUploader,
+        didUpdateWithProgress progress: BlobUploadProgress
+    )
+    func uploader(_ uploader: BlobStreamUploader, didFinishWithProgress progress: BlobUploadProgress)
+    func uploader(_ uploader: BlobStreamUploader, didFailWithError error: Error)
+}
+
+/// Class used to upload block blobs.
+public class BlobStreamUploader {
+    // MARK: Properties
+
+    public weak var delegate: BlobUploadDelegate?
+
+    /// Location of the blob to be uploaded on the device
+    public let uploadSource: URL
+
+    /// Name of the blob being uploaded
+    public let blobName: String
+
+    /// Name of the container containing the blob
+    public let containerName: String
+
+    /// Properties of the blob being uploaded.
+    public var blobProperties: BlobProperties?
+
+    /// The total bytes uploaded.
+    public var progress = 0
+
+    /// The list of blocks for the blob upload.
+    public var blockList = [(range: Range<Int>, blockId: String)]()
+
+    /// Internal list that maps the order of block IDs
+    internal var blockIdMap = [String: Int]()
+
+    /// Logs completed block IDs and the order they *should* be in
+    internal var completedBlockMap = [String: Int]()
+
+    /// Indicates if the upload  is complete.
+    public var isComplete: Bool {
+        return progress == fileSize
+    }
+
+    /// Indicates if the upload  is encrypted.
+    public var isEncrypted: Bool {
+        return options.encryptionOptions?.key != nil || options.encryptionOptions?.keyResolver != nil
+    }
+
+    internal let client: StorageBlobClient
+
+    internal let options: UploadBlobOptions
+
+    /// Size, in bytes, of the file to be uploaded.
+    internal var fileSize: Int!
+
+    // MARK: Initializers
+
+    /// Create a `BlobStreamUploader` object.
+    /// - Parameters:
+    ///   - client: A`StorageBlobClient` reference.
+    ///   - name: The name of the blob to upload.
+    ///   - container: The name of the container the blob is contained in.
+    ///   - options: An `UploadBlobOptions` object to control the upload process.
+    public init(
+        client: StorageBlobClient,
+        delegate: BlobUploadDelegate? = nil,
+        source: URL,
+        name: String,
+        container: String,
+        properties: BlobProperties? = nil,
+        options: UploadBlobOptions? = nil
+    ) throws {
+        let manager = FileManager.default
+        guard manager.fileExists(atPath: source.path) else {
+            throw AzureError.general("File not found: \(source.path)")
```

Comment:
It occurs to me we should probably make some Storage-specific error subclasses.

Reply:
I added a fileSystem error type under AzureError. I did a dedicated pass to replace `AzureError.general` with more appropriate categories. We should discuss this as part of our public API review.